### PR TITLE
MastForest maximum node length invariant

### LINKED ISSUE #1392: `MastForest`: Ensure that the number of nodes is less than 2^30
The serialization format of `MastForest` introduced in #1370 implicitly assumes that the internal value of a `MastNodeId` fits in 30 bits. Hence, `MastForest` should enforce this constraint.

### PR DESCRIPTION
**Pull request by sergerad**
Closes #1392

* Enforce invariant that MastForest node length is < 2^30
* Return errors from add_node() and enforce_node()
* Map errors to AssemblyError

### CODE REVIEW

**Comment by sergerad** [comment]
@plafer LMK if this makes sense.

And are we happy panicking in `add_node()`? Or do we want to return error and refactor usage?

**Comment by plafer** [comment]
@sergerad Thank you for taking this on, it is super appreciated!

The current implementation looks great; however, could you rebase it on the `plafer-mast-forest-serialization` branch? We can then wait for that PR to be merged before we merge this one. When I created #1392, I assumed the PR would be merged by the time anyone started working on it 😅.

Hence, the current implementation can stay the same, and the new `MastNodeId::from_u32_safe()` could be modified to use the new `TryFrom` that you implemented.

> And are we happy panicking in add_node()? Or do we want to return error and refactor usage?

If you are willing to do the work, returning an error would be much better as it's definitely the direction we want to go in. This would also require a similar change to `MastForestBuilder::ensure_node()`.

**Comment by plafer** [comment]
Actually, thinking about it a bit more, I think it makes more sense for `MastForest` to know about the $2^{30}$ limit on number of nodes as opposed to the `MastNodeId` (a constraint that ultimately comes from the `MastForest` serialization format). So:

- `MastNodeId::from_u32_safe()` remains unchanged
- `MastForest::add_node()` contains the $2^{30}$ node limit check, and returns an error instead of panicking

**Comment by sergerad** [comment]
Thanks very much for that guidance @plafer. I will make those changes including the removal of the panic.

**Comment by sergerad** [comment]
@plafer I have added the node count invariant logic to MastForest and updated the relevant fns to return errors.

Most of the diffs are due to adding unwrap() in test files. All other files should be handling the new errors, not unwrapping.

I have done my best to handle those errors appropriately and map to AssemblyError. Let me know if and how to change the approach I have there with MastForestError.

File: core/src/mast/serialization/mod.rs
Comment (bobbinth) [inline core/src/mast/serialization/mod.rs:125]: The only reason this could fail is because we've serialized more than 2^30 nodes, right? If so, I'd probably use the `InvalidValue` error variant.

File: assembly/src/assembler/mod.rs
Comment (bobbinth) [inline assembly/src/assembler/mod.rs:799]: Is `.map_err()` here (and in other similar places) necessary? I would have thought that the compiler would be able to infer the conversion automatically and just `mast_forest_builder.ensure_node(split_node)?` would have been enough.

File: assembly/src/assembler/instruction/procedures.rs
Comment (bobbinth) [inline assembly/src/assembler/instruction/procedures.rs:101]: I would probably write this as:
```rust
match mast_forest_builder.find_procedure_root(mast_root) {
    None => {
        // If the MAST root called isn't known to us, make it an external
        // reference.
        let external_node = MastNode::new_external(mast_root);
        mast_forest_builder.ensure_node(external_node)?
    }
    Some(root) => root
}
```
To keep it consistent with how we have things below.

File: assembly/src/assembler/basic_block_builder.rs
Comment (bobbinth) [inline assembly/src/assembler/basic_block_builder.rs:138]: Shouldn't this `unwrap()` also be an error? It does complicate the return type a bit - but I'm not sure if there is a good way to get around this.

**Review by bobbinth — COMMENTED** [review]
Looks good! Thank you for tackling this! I left a few small comments inline.

File: core/src/mast/serialization/mod.rs
Comment (plafer) [inline core/src/mast/serialization/mod.rs:125]: Let's use `DeserializationError::InvalidValue` instead, and maybe with an error string that looks like `format!("failed to add node to MAST forest while deserializing: {}", e.to_string())`

**Review by plafer — COMMENTED** [review]
Looks good! Just left a few minor comments

File: assembly/src/assembler/basic_block_builder.rs
Comment (sergerad) [reply assembly/src/assembler/basic_block_builder.rs:138]: Good catch, I missed this!

File: core/src/mast/mod.rs
Comment (sergerad) [reply core/src/mast/mod.rs:114]: Yes that's right - `1 << 30` is the non inclusive upper bound, which doesn't make sense as a value for `MAX_NODES`

File: assembly/src/assembler/basic_block_builder.rs
Comment (sergerad) [reply assembly/src/assembler/basic_block_builder.rs:138]: Yep, missed this one - treated it as a test file

File: assembly/src/assembler/mod.rs
Comment (sergerad) [reply assembly/src/assembler/mod.rs]: Updated. Just noting that the conversion required MastForestError to implement the miette::Diagnostic trait to for this. So the core crate now imports miette.

**Comment by sergerad** [comment]
Think I have covered all the current comments. Thanks!

**Review by plafer — APPROVED** [review]
LGTM! (let's fix clippy)

**Comment by plafer** [comment]
Merged #1370, and changed the target branch to `next`

**Comment by sergerad** [comment]
Fixed clippy and changelog 👍

**Review by plafer — APPROVED** [review]
🙏

File: assembly/src/assembler/mod.rs
Comment (bobbinth) [reply assembly/src/assembler/mod.rs]: Ah! In that case, I'd probably prefer to keep the extra `map_err()` here and avoid an additional import.

File: core/Cargo.toml
Comment (bobbinth) [inline core/Cargo.toml:1]: I think addressing https://github.com/0xPolygonMiden/miden-vm/pull/1394#discussion_r1683579436 should also revert changes to this file.

**Review by bobbinth — COMMENTED** [review]
Looks good! Thank you! I left one small comment inline.

**Review by bobbinth — APPROVED** [review]
All looks good! Thank you!

**Pull request merged**
